server/asset/btc/rpcclient.go

```diff
@@ -130,8 +164,195 @@ func (rc *RPCClient) GetRawTransactionVerbose(txHash *chainhash.Hash) (*btcjson.
+
+		if txCount >= 101 {
+			break
+		}
+
+		// Not enough transactions to count yet.
+		verboseBlock, err := rc.GetBlockVerbose(blockHash)
+		if err != nil {
+			return 0, err
+		}
+
+		blockHash, err = chainhash.NewHashFromStr(verboseBlock.PreviousHash)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// rounded average
+	return uint64(math.Round(float64(weight) / float64(txCount))), nil
+}
+
+// medianFeesTheHardWay calculates the median fees from the previous block(s).
+// medianFeesTheHardWay is used for assets that don't have a getblockstats RPC,
+// and is only useful for non-segwit assets.
+func (rc *RPCClient) medianFeesTheHardWay() (uint64, error) {
+	const numTxs = 101
+
+	iHash, err := rc.GetBestBlockHash()
+	if err != nil {
+		return 0, err
+	}
+
+	txs := make([]*wire.MsgTx, 0, numTxs)
+	prevOuts := make(map[chainhash.Hash]map[int]int64, numTxs)
+
+	var blocksChecked int
+
+out:
+	for len(txs) < numTxs {
+		blocksChecked++
+		if blocksChecked > rc.maxFeeBlocks {
+			return 0, ErrNoCompetition
+		}
+
+		blk, err := rc.GetBlockVerbose(iHash)
+		if err != nil {
+			return 0, err
+		}
+
+		if len(blk.Tx) == 0 {
+			return 0, fmt.Errorf("no transactions?")
+		}
+
+		rawTxs := blk.Tx[1:] // skip coinbase
+		rand.Shuffle(len(rawTxs), func(i, j int) { rawTxs[i], rawTxs[j] = rawTxs[j], rawTxs[i] })
+
+		for _, txid := range rawTxs {
+			txHash, err := chainhash.NewHashFromStr(txid)
+			if err != nil {
+				return 0, err
+			}
+
+			utilTx, err := rc.GetRawTransaction(txHash)
+			if err != nil {
+				return 0, err
+			}
+
+			tx := utilTx.MsgTx()
+
+			for _, vin := range tx.TxIn {
+				prevOut := vin.PreviousOutPoint
+				prevs := prevOuts[prevOut.Hash]
+				if len(prevs) == 0 {
+					prevs = make(map[int]int64, 1)
+					prevOuts[prevOut.Hash] = prevs
+				}
+				prevs[int(prevOut.Index)] = 0 // placeholder
+			}
+			txs = append(txs, tx)
+
+			if len(txs) >= numTxs {
+				break out
+			}
+		}
+
+		iHash, err = chainhash.NewHashFromStr(blk.PreviousHash)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Fetch all the previous outpoints and log the values.
+	for txHash, prevs := range prevOuts {
+		utilTx, err := rc.GetRawTransaction(&txHash)
+		if err != nil {
+			return 0, fmt.Errorf("GetRawTransaction error: %v", err)
+		}
+		tx := utilTx.MsgTx()
+		for vout := range prevs {
+			if len(tx.TxOut) < vout+1 {
+				return 0, fmt.Errorf("too few outputs")
+			}
+			prevs[vout] = tx.TxOut[vout].Value
+		}
+	}
+
+	// Do math.
+	rates := make([]uint64, 0, numTxs)
+	for _, tx := range txs {
+		var in, out int64
+		for _, vin := range tx.TxIn {
+			prevOut := vin.PreviousOutPoint
+			in += prevOuts[prevOut.Hash][int(prevOut.Index)]
+		}
+		for _, vout := range tx.TxOut {
+			out += vout.Value
+		}
+		fees := in - out
+		if fees < 0 {
+			return 0, fmt.Errorf("fees < 0 for tx %s", tx.TxHash())
+		}
+		sz := tx.SerializeSize()
+		if sz == 0 {
+			return 0, fmt.Errorf("size 0 tx %s", tx.TxHash())
+		}
+		rates = append(rates, uint64(fees)/uint64(sz))
```

Comment:
Under non-error conditions you will always append to length `numTxs`, so I'd be tempted to do `for i, tx := range txs` and `rates[i] = ...`  to spare modifying len in the slice header on each iteration, but nbd.